File: Solution/【017】电话号码的字母组合/电话号码的字母组合.cpp
/*
（1）注意 map 初始化赋值的方式
（2）此处核心思想是：先添加一个 字符串，然后把这个字符串加入新的字符后删除前面的字符串
*/




class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        if(digits.empty() || digits.size() == 0){
            return result;
        }
        map<char, string> my_map = {{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
        result.push_back("");
        for(int i = 0; i < digits.size() ; i ++){
            char num = digits.at(i);
            if(num == '1'){
                return result;
            }
            int len = result.size();
            for(int j = 0 ; j < len ; j ++){
                for (auto ch: my_map[num]){   //遍历按键对应字符
                    result.push_back(result[j] + ch);
                }
            }
            result.erase(result.begin(),result.begin()+len);
        }
        return result;
    }
};

File: Solution/【017】电话号码的字母组合/电话号码的字母组合.java
/*
（1）先用 map 存储好每个数字字符对应的字符串
（2）回溯，注意 for 循环中遍历的是 每个数字字符对应的字符串中每个字符
*/





class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new LinkedList<String>();
        if (digits == null || digits.length() == 0) {
            return result;
        }
        Map<Character, String> map = new HashMap<Character, String>();
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");
        dfs("", digits, 0, result, map);
        return result;
    }
    public void dfs(String s, String digits, int i, List<String> result, Map<Character, String> map) {
        if (i == digits.length()) {
            result.add(s);
            return;
        }
        String letters = map.get(digits.charAt(i));
        for (int j = 0; j < letters.length(); j++){
            dfs(s+letters.charAt(j),digits,i+1,result,map);
        }
    }
}